src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs

```diff
@@ -1082,15 +1082,45 @@ public bool TakeStartLine(ReadableBuffer buffer, out ReadCursor consumed, out Re
 
             var needDecode = false;
             var pathBegin = methodEnd + 1;
-            var pathToEndSpan = span.Slice(pathBegin, span.Length - pathBegin);
-            pathBegin = 0;
+            var spaceIndex = -1;
+            var questionMarkIndex = -1;
+            var percentageIndex = -1;
+            var pathEnd = -1;
 
-            // TODO: IndexOfAny
-            var spaceIndex = pathToEndSpan.IndexOf(ByteSpace);
-            var questionMarkIndex = pathToEndSpan.IndexOf(ByteQuestionMark);
-            var percentageIndex = pathToEndSpan.IndexOf(BytePercentage);
+            // TODO: State machineify
+            for (int i = pathBegin; i < span.Length; i++)
```

Comment:
I'm looping over valid indexes up to `span.Length`. `pathBegin` can't be > span.Length since span is 0 based. Unlike ArraySegment, we don't need to add the offset.